### PR TITLE
feat(settings modal)

### PR DESCRIPTION
**Pull request by HugoDF**
One for @KevinBatdorf would be great to get some design feedback from @stephenoldham as well

- add "info" icon + modal that opens with information
- for now I've disabled it on production builds since we don't have much content for it

Demo:

Kapture 2020-12-05 at 17 56 36

### CODE REVIEW

**Comment by KevinBatdorf** [comment]
This feels a little bit off to me. How would you feel about a more general settings area?

modal3


I kind of agree with you that we shouldn't need to have document features outright.  Instead, we could disguise documentation as settings. If we decide allowing something like `x-devtools-ignore` we could instead include a settings area with something like 

```html
Enable advanced visibility mode
<small>This lets you hide or show components based on a specific selector you define.</small>
<input x-modal="ignoreSelector">
```

**Comment by HugoDF** [comment]
> This feels a little bit off to me. How would you feel about a more general settings area?
> 
> modal3
> 
> 
> I kind of agree with you that we shouldn't need to have document features outright.  Instead, we could disguise documentation as settings. If we decide allowing something like `x-devtools-ignore` we could instead include a settings area with something like 
> 
> ```html
> Enable advanced visibility mode
> <small>This lets you hide or show components based on a specific selector you define.</small>
> <input x-modal="ignoreSelector">
> ```

Looks good to me 👍

As long as the users can learn about it through the UI it's all good.

Shall I close this or do you want to push those changes to this PR?

**Comment by KevinBatdorf** [comment]
I can push here. It's just boilerplate stuff though at the moment. Could we easily extract the panel.html file into separate files to manage it a little better? What would you suggest?

**Comment by KevinBatdorf** [comment]
I pushed the files here. Take a look and let me know if you want anything changed or removed

**Comment by HugoDF** [comment]
> I can push here. It's just boilerplate stuff though at the moment. Could we easily extract the panel.html file into separate files to manage it a little better? What would you suggest?

🤔 there's a couple of options

1) We could switch to a templating system that allows partials, maybe nunjucks or https://edge.adonisjs.com/ (which has a blade-like syntax so I assume we could do partials)

We would need to mess around with the build setup in order to do that, probably by pre-processing the HTML 🤔🤔 pretty much removing the HTML from rollup copy step and running the templating library/some static site generator instead or creating a custom rollup plugin

2) we could import HTML partials into the JS file (devtools.js) as strings (using rollup + rollup plugin HTML) and then use `x-html` to inject the markup

3) we could split some of the markup into HTML partials, load them using `fetch` and inject using `x-html`

None of the 3 are ideal, 1) is best for performance (HTML would be fully generated at build time) but 2) and 3) will probably be easier to set up.

File: packages/shell-chrome/assets/panel.html [hunk 1]
```diff
@@ -395,9 +411,59 @@ <h5 class="flex items-center px-2 leading-7 font-mono whitespace-no-wrap">
                         </div>
                     </div>
                 </div>
+                <div
+                    x-show="settingsPanelEnabled && settingsPanelOpen"
+                    class="inset-0 absolute flex flex-col items-center justify-center"
+                    style="z-index: 2147483647"
```
Comment (HugoDF) [inline packages/shell-chrome/assets/panel.html:417]: We can use a tailwind class for this I think

File: packages/shell-chrome/assets/panel.html [hunk 2]
```diff
@@ -395,9 +411,59 @@ <h5 class="flex items-center px-2 leading-7 font-mono whitespace-no-wrap">
...
+                                    <h2 class="text-xl font-light">Settings</h2>
+                                    <nav class="flex flex-col items-start text-ms mt-3 text-cool-gray-500 space-y-1">
+                                        <button>Prefrences</button>
+                                        <button>Workspace</button>
+                                        <button>Experiments</button>
+                                    </nav>
+                                </div>
+                                <div class="flex-grow">
+                                    <div class="w-full border-b border-cool-gray-200 mb-6 pb-2">
+                                        <h3 class="text-2xl font-light">Prefrences</h3>
```
Comment (HugoDF) [inline packages/shell-chrome/assets/panel.html:457]:
```suggestion
                                        <h3 class="text-2xl font-light">Preferences</h3>
```

File: packages/shell-chrome/assets/panel.html [hunk 3]
```diff
@@ -395,9 +411,59 @@ <h5 class="flex items-center px-2 leading-7 font-mono whitespace-no-wrap">
...
+                                        d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
+                                        clip-rule="evenodd"
+                                    />
+                                </svg>
+                            </button>
+                            <div class="flex space-x-12">
+                                <div>
+                                    <h2 class="text-xl font-light">Settings</h2>
+                                    <nav class="flex flex-col items-start text-ms mt-3 text-cool-gray-500 space-y-1">
+                                        <button>Prefrences</button>
```
Comment (HugoDF) [inline packages/shell-chrome/assets/panel.html:450]:
```suggestion
                                        <button>Preferences</button>
```

**Review by HugoDF — COMMENTED** [review]
What's up with the massive package-lock diff?

Otherwise looks fine to me

**Comment by KevinBatdorf** [comment]
> What's up with the massive package-lock diff?

I ran `npm i` to pull in that divider bar. I can remove it though

**Comment by HugoDF** [comment]
> > What's up with the massive package-lock diff?
> 
> I ran `npm i` to pull in that divider bar. I can remove it though

Weird, that didn't happen when I did the npm install

**Comment by HugoDF** [comment]
Otherwise, let's add a Cypress test for show/hide of the modal, merge this and we can edit the modal content in #109 ?

**Comment by KevinBatdorf** [comment]
> Otherwise, let's add a Cypress test for show/hide of the modal, merge this and we can edit the modal content in #109 ?

I can add one now. If I can get it working. Maybe I just change the port number? Does it matter much to you?

> 1. is best for performance

I say let's go with #1 then. Do you want me to set this up? Might not be until the weekend though. Want to use https://edge.adonisjs.com/ ?

**Comment by HugoDF** [comment]
@KevinBatdorf not too fussed if I'm honest, I think we can add some JS to the rollup config that will do the template compilation

But let's keep it as a separate bit of functionality, it's not technically a blocker for anything, just nicer development experience as the panel gets larger

**Comment by KevinBatdorf** [comment]
If that's more convenient. I think having it clean and separate early on will keep it from getting unwieldy fast. Especially if we plan to start adding more features like event monitoring, Spruce integration, etc

**Review by KevinBatdorf — APPROVED** [review]

File: cypress/integration/settings.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+it('should be able to open and close the settings modal', () => {
+    cy.visit('/')
+        .get('[data-testid=component-name]')
+        .get('#settingsPanelOpenButton')
+        .click()
+        .get('#settingsPanel')
```
Comment (HugoDF) [inline cypress/integration/settings.js:6]: Could we use `data-testid`'s instead? (Unless we're using the IDs for something)

File: cypress/integration/settings.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+it('should be able to open and close the settings modal', () => {
+    cy.visit('/')
+        .get('[data-testid=component-name]')
+        .get('#settingsPanelOpenButton')
```
Comment (HugoDF) [inline cypress/integration/settings.js:4]: Might be cleaner to wait for the setting button to appear
```suggestion
        .get('#settingsPanelOpenButton')
        .should('be.visible')
```

**Review by HugoDF — COMMENTED** [review]
Last few comments

You're right about the pre-processing, we can create an issue to make it more visible/easy to track/discuss

**Comment by KevinBatdorf** [comment]
Can you check it again? I changed the IDs and updated the test

**Comment by HugoDF** [comment]
Nice one @KevinBatdorf

**Pull request merged**